### PR TITLE
gh-143008: fix TextIOWrapper.truncate via re-entrant flush

### PR DESCRIPTION
**Pull request by yihong0618**
This patch fix all  re-entrant flush problem 
by adding an attr flushing
and. refactor the file check to `_textiowrapper_flush`

* Issue: gh-143008

all crash problem in 143008 can be fix cc @jackfromeast can you help to check?
thank you very much!

### CODE REVIEW

File: Lib/test/test_io/test_textio.py [hunk 1]
```diff
@@ -1560,6 +1560,54 @@ def closed(self):
         wrapper = self.TextIOWrapper(raw)
         wrapper.close()  # should not crash
 
+    def test_reentrant_detach_during_flush(self):
+        # gh-143008: Reentrant detach() during flush should raise RuntimeError
+        # instead of crashing.
+        wrapper = None
+
+        class BadRaw(self.RawIOBase):
```
Comment (cmaloney) [inline Lib/test/test_io/test_textio.py]: There's a number of mock helpers defined in `test_io/utils.py` to make it so can reduce this to just the methods that matter. I think `self.MockRawIO` will work here

File: Lib/test/test_io/test_textio.py [hunk 2]
```diff
@@ -1560,6 +1560,54 @@ def closed(self):
...
+            ('detach', lambda: wrapper.detach()),
+            ('seek', lambda: wrapper.seek(0)),
+            ('tell', lambda: wrapper.tell()),
+            ('reconfigure', lambda: wrapper.reconfigure(line_buffering=True)),
+        ]
+        for name, method in tests:
+            with self.subTest(name):
+                wrapper = self.TextIOWrapper(EvilBuffer(BadRaw()), encoding='utf-8')
+                self.assertRaisesRegex(RuntimeError, "reentrant", method)
+                wrapper = None
```
Comment (cmaloney) [inline Lib/test/test_io/test_textio.py]: can we make the `gc` / closing here more explicit? `del wrapper`?

File: Misc/NEWS.d/next/Library/2025-12-21-17-56-37.gh-issue-143008.aakErJ.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix crash in :class:`io.TextIOWrapper` when reentrant ``detach()`` called.
```
Comment (cmaloney) [inline Misc/NEWS.d/next/Library/2025-12-21-17-56-37.gh-issue-143008.aakErJ.rst]:
```suggestion
Fix crash in :class:`io.TextIOWrapper` when reentrant :meth:`io.TextIOBase.detach` called.
```

File: Modules/_io/textio.c [hunk 4]
```diff
@@ -725,6 +726,16 @@ struct textio
 
 #define textio_CAST(op) ((textio *)(op))
 
+/* gh-143007 need to check for reentrant flush */
+static inline int
+_textiowrapper_flush(textio *self)
+{
+    self->flushing = 1;
```
Comment (cmaloney) [inline Modules/_io/textio.c]: Why can't this be in `_io_TextIOWrapper_flush_impl`?

File: Modules/_io/textio.c [hunk 5]
```diff
@@ -1636,9 +1653,11 @@ _textiowrapper_writeflush(textio *self)
     Py_DECREF(pending);
 
     PyObject *ret;
+    self->flushing = 1;
     do {
         ret = PyObject_CallMethodOneArg(self->buffer, &_Py_ID(write), b);
```
Comment (cmaloney) [inline Modules/_io/textio.c:1669]: every loop iteration / `write` call could execute arbitrary interpreter code that could invalidate `self->buffer` so likely need to check on each loop iteration.

**Review by cmaloney — COMMENTED** [review]
I don't like adding a new `flushing` variable here. To me it doesn't solves the core issue: Code assumes `self->buffer` is non-NULL because of checks which become invalid by the time it's actually dereferenced/used ([TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) general bug class). I think `TextIOWrapper` already has the information we need to be looking at / checking for (`CHECK_ATTACHED` macro / `self->detached`, `self->buffer == NULL`, and `self->ok`, ...). Those seem to be tracking very close to the same thing and should be usable to solve this group of issues.

I think it would be better here to use `CHECK_ATTACHED` just before `self->buffer` is used / passed to a call / returned; making sure there aren't any other calls to interpreter methods betwen the `CHECK_ATTACHED` + `self->buffer` usage / dereference.

I suspect `BufferedIO` will need a similar set of fixes: It also has a concept of an underlying stream that can be "detached" and which user code may detach during various operations.

Some comments on specifics inline, the test_textio ones I think are important in further changes, implementation details less so if not adding/using the `flushing` member.

**Comment by yihong0618** [comment]
> I don't like adding a new `flushing` variable here. To me it doesn't solves the core issue: Code assumes `self->buffer` is non-NULL because of checks which become invalid by the time it's actually dereferenced/used ([TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) general bug class). I think `TextIOWrapper` already has the information we need to be looking at / checking for (`CHECK_ATTACHED` macro / `self->detached`, `self->buffer == NULL`, and `self->ok`, ...). Those seem to be tracking very close to the same thing and should be usable to solve this group of issues.
> 
> I think it would be better here to use `CHECK_ATTACHED` just before `self->buffer` is used / passed to a call / returned; making sure there aren't any other calls to interpreter methods betwen the `CHECK_ATTACHED` + `self->buffer` usage / dereference.
> 
> I suspect `BufferedIO` will need a similar set of fixes: It also has a concept of an underlying stream that can be "detached" and which user code may detach during various operations.
> 
> Some comments on specifics inline, the test_textio ones I think are important in further changes, implementation details less so if not adding/using the `flushing` member.

I think you are right but `CHECK_ATTACHED` in the middle of code is not a good name and easy to forget add it.

**Comment by yihong0618** [comment]
And I can change it to my first solution or waiting for other discuss here?

File: Lib/test/test_io/test_textio.py (hunk 2 above)
Comment (yihong0618) [reply Lib/test/test_io/test_textio.py]: thanks

File: Lib/test/test_io/test_textio.py (hunk 1 above)
Comment (yihong0618) [reply Lib/test/test_io/test_textio.py]: thanks

File: Modules/_io/textio.c (hunk 5 above)
Comment (yihong0618) [reply Modules/_io/textio.c:1669]: will check, but I think maybe not need

File: Modules/_io/textio.c (hunk 4 above)
Comment (yihong0618) [reply Modules/_io/textio.c]: it can, the design is cover it in _PyFile_Flush

**Comment by cmaloney** [comment]
I agree `CHECK_ATTACHED` is easy to forget and I'd argue this bug is the result of some cases missing it.. It is a relatively small change though and matches the rest of the `textio.c` code style / convention. I have been working towards changing `textio.c` more generally, but that's too big for a likely back-ported bugfix.

Can you explain what your first solution is? I'm not understanding what you're referring to.

**Comment by yihong0618** [comment]
> I agree `CHECK_ATTACHED` is easy to forget and I'd argue this bug is the result of some cases missing it.. It is a relatively small change though and matches the rest of the `textio.c` code style / convention. I have been working towards changing `textio.c` more generally, but that's too big for a likely back-ported bugfix.
> 
> 
> 
> Can you explain what your first solution is? I'm not understanding what you're referring to.

https://github.com/python/cpython/issues/143008#issuecomment-3677398247

**Review by serhiy-storchaka — COMMENTED** [review]
This will break the case when `flush()` was intentionally overridden.

I am not also sure that it fixes the core issue, it just makes a particular reproducer no longer working.

**Comment by yihong0618** [comment]
> This will break the case when `flush()` was intentionally overridden.
> 
> I am not also sure that it fixes the core issue, it just makes a particular reproducer no longer working.

will dig it, thank you

**Comment by yihong0618** [comment]
> I agree `CHECK_ATTACHED` is easy to forget and I'd argue this bug is the result of some cases missing it.. It is a relatively small change though and matches the rest of the `textio.c` code style / convention. I have been working towards changing `textio.c` more generally, but that's too big for a likely back-ported bugfix.
> 
> Can you explain what your first solution is? I'm not understanding what you're referring to.

after some check and work I do not think `CHECK_ATTACHED`  before every is a good design will dig more
maybe  I can find some better solution to fix all of them, if I failed will close this waiting for some expert to finish
this, thank you very much for review

File: Modules/_io/textio.c (hunk 4 above)
Comment (yihong0618) [reply Modules/_io/textio.c]: dropped

File: Modules/_io/textio.c (hunk 5 above)
Comment (yihong0618) [reply Modules/_io/textio.c:1669]: dropped

**Comment by vstinner** [comment]
https://github.com/python/cpython/pull/145957 was merged instead. Thanks for your contribution anyway! The merged change is based on your work and you're credited in the change ;-)
